### PR TITLE
Add export_to_tensorboard for auto_grad/profiler

### PR DESCRIPTION
**Pull request by guyang3532**

### CODE REVIEW

File: torch/autograd/profiler.py [hunk 1]
```diff
@@ -495,6 +495,23 @@ def export_chrome_trace(self, path):
             return self.function_events.export_chrome_trace(path)
     export_chrome_trace.__doc__ = EventList.export_chrome_trace.__doc__
 
+    def export_to_tensorboard(self, log_dir, run_name=None, worker_num=0):
+        import os, time, socket
+        if not os.path.exists(log_dir):
+            os.mkdir(log_dir)
+        elif not os.path.isdir(log_dir):
+            raise RuntimeError(log_dir + " is not a directory.")
```
Comment (gaoteng-git) [inline torch/autograd/profiler.py:503]: "{} is not a directory.".format(log_dir)
Same for other strings below

File: torch/autograd/profiler.py [hunk 2]
```diff
@@ -495,6 +495,23 @@ def export_chrome_trace(self, path):
...
+
+        if run_name == None:
```
Comment (gaoteng-git) [inline torch/autograd/profiler.py:505]: Maybe "run_name is None" is better

File: torch/autograd/profiler.py [hunk 3]
```diff
@@ -495,6 +495,23 @@ def export_chrome_trace(self, path):
             return self.function_events.export_chrome_trace(path)
     export_chrome_trace.__doc__ = EventList.export_chrome_trace.__doc__
 
+    def export_to_tensorboard(self, log_dir, run_name=None, worker_num=0):
+        import os, time, socket
+        if not os.path.exists(log_dir):
+            os.mkdir(log_dir)
```
Comment (gaoteng-git) [inline torch/autograd/profiler.py:501]: Need use try-catch?
        try:
            os.mkdir(logdir)
        except os.OSError as e:
            # throw exception?

**Review by ilia-cher — CHANGES_REQUESTED** [review]
the added function named export_to_tensorboard, but it just simply calls export_chrome_trace, that is confusing and unnecessary;

as discussed, the way to implement this is a) work with new profiler in torch/profiler b) pass format parameter in save trace (kineto api) c) add support for multiple formats into kineto cc. @gdankel

File: torch/autograd/profiler.py [hunk 4]
```diff
@@ -495,6 +495,20 @@ def export_chrome_trace(self, path):
...
+    def export_to_tensorboard(self, dir_name, prefix=None):
+        import os, time, socket
+        if not os.path.isdir(dir_name):
+            try:
+                os.mkdir(dir_name)
+            except Exception as e:
+                raise RuntimeError("Can't create directory: " + dir_name)
+
+        file_name = socket.gethostname() + "_" + str(os.getpid()) + "_" + \
+                    str(int(time.time()*1000)) + ".pt.trace.json"
```
Comment (nbcsm) [inline torch/autograd/profiler.py:507]: try to use `format`

File: torch/autograd/profiler.py [hunk 5]
```diff
@@ -495,6 +495,20 @@ def export_chrome_trace(self, path):
...
+        if prefix:
+            file_name = prefix + "_" + file_name
```
Comment (nbcsm) [inline torch/autograd/profiler.py:509]: `hostname_pid.timestamp.pt.trace.json` or `prefix.timestamp.pt.trace.json`

File: torch/autograd/profiler.py [hunk 6]
```diff
@@ -495,6 +495,20 @@ def export_chrome_trace(self, path):
             return self.function_events.export_chrome_trace(path)
     export_chrome_trace.__doc__ = EventList.export_chrome_trace.__doc__
 
+    def export_to_tensorboard(self, dir_name, prefix=None):
```
Comment (nbcsm) [inline torch/autograd/profiler.py:498]: need a test

**Comment by guyang3532** [comment]
This helps users easy to formalize directory structure and file names for tensorboard plugin especially in distributed cases.

For DDP example:

    import torch.multiprocessing as mp
    from torch.nn.parallel import DistributedDataParallel as DDP

    def func(rank, world_size):
        with torch.autograd.profiler.profile(use_cuda=True) as prof:
            ddp_model = DDP(model, device_ids=[rank])
            ...
            loss.backward()
            optimizer.step()
	    prof.export_to_tensorboard()
	    # This will automatically output to 8 files for 8 gpus respectively.

    gpu_num = 8
    mp.spawn(func,args=(gpu_num,),nprocs=gpu_num)

In this case, users only need to call p.export_to_tensorboard(), then the result files will aotomatically output to different files for each gpu.
Users need not to known which node and which gpu is each output file for.
And then the tensorboard plugin api can easy to recognize and parse them.

This function will be called from the new profiler api in torch/profiler when that pr be merged.
@ilia-cher

**Comment by ilia-cher** [comment]
let's continue the discussion after https://github.com/pytorch/pytorch/pull/48280 is landed

**Comment by ilia-cher** [comment]
Hi, now that #48280 is landed, this kind of functionality is supported by specifying `on_trace_ready` parameter - you're free to pass any lambda function, e.g. the one that's generated from the logging directory path. Also let's keep the profiler code changes in torch/autograd/profiler.py to a minimum, as we're transitioning away from the older API.

**Comment by guyang3532** [comment]
resubmit it at https://github.com/pytorch/pytorch/pull/50875
